pkg/pillar/attest/attest.go

```diff
@@ -0,0 +1,346 @@
+	case EventNoQuoteCertRecvd:
+		return "EventNoQuoteCertRecvd"
+	case EventAttestEscrowRecorded:
+		return "EventAttestEscrowRecorded"
+	case EventAttestEscrowFailed:
+		return "EventAttestEscrowFailed"
+	case EventAttestSuccessful:
+		return "EventAttestSuccessful"
+	case EventRestart:
+		return "EventRestart"
+	default:
+		return "Unknown Event"
+	}
+}
+
+//String returns human readable string of a State
+func (state State) String() string {
+	switch state {
+	case StateNone:
+		return "StateNone"
+	case StateNonceWait:
+		return "StateNonceWait"
+	case StateInternalQuoteWait:
+		return "StateInternalQuoteWait"
+	case StateAttestWait:
+		return "StateAttestWait"
+	case StateAttestEscrowWait:
+		return "StateAttestEscrowWait"
+	case StateRestartWait:
+		return "StateRestartWait"
+	case StateComplete:
+		return "StateComplete"
+	default:
+		return "Unknown State"
+	}
+}
+
+//Verifier needs to be implemented by the consumer of this package
+//It contains interface definitions for interacting with attestation server
+type Verifier interface {
+	SendNonceRequest(ctx *Context) error
+	SendAttestQuote(ctx *Context) error
+	SendAttestEscrow(ctx *Context) error
+}
+
+//TpmAgent needs to be implemented by the consumer of this package
+//It contains interface definitions for interacting with TPM manager
+type TpmAgent interface {
+	SendInternalQuoteRequest(ctx *Context) error
+}
+
+//Various error codes to be returned to this package from external interfaces
+var (
+	ErrControllerUnavailable = errors.New("Controller is unavailable")
+	ErrNonceMismatch         = errors.New("Nonce Mismatch")
+	ErrQuoteMismatch         = errors.New("Quote Mismatch")
+	ErrNoCertYet             = errors.New("No Cert Found")
+	ErrInfoTokenInvalid      = errors.New("Provided Integrity Token is Invalid")
+	ErrTpmAgentUnavailable   = errors.New("TPM agent is unavailable")
+)
+
+//External interfaces to the state machine
+//For unit-testing, these will be redirected to their mock versions.
+var tpmAgent TpmAgent
+var verifier Verifier
+
+//Context has all the runtime context required to run this state machine
+type Context struct {
+	event                 Event
+	state                 State
+	restartTimer          *time.Timer
+	eventTrigger          chan Event
+	retryTime             time.Duration
+	restartRequestPending bool
+}
+
+//Transition represents an event triggered from a state
+type Transition struct {
+	event Event
+	state State
+}
+
+//EventHandler represents a handler function for a Transition
+type EventHandler func(*Context) error
+
+//the state machine
+var transitions = map[Transition]EventHandler{
+	{EventInitialize, StateNone}:                       handleInitializeWhlNone,                       //goes to NonceWait
+	{EventRetryTimerExpiry, StateRestartWait}:          handleRetryTimerExpiryWhlRestartWait,          //goes to NonceWait
+	{EventRestart, StateRestartWait}:                   handleRestartWhlStateInProgress,               //goes to RestartWait
```

Comment:
There isn't a state called "InProgress".
Hmm, I wonder how much commonality there will be for the individual handlers vs. having generic handlers like InProgress, which might end up looking at state as part of their logic.

If the key event handlers end up being generic InProgress ones, then it might make more sense to define a table with handlers key'ed by just the event, and then have the code look at the current state to determine what to do.

Reply:
> There isn't a state called "InProgress".
> Hmm, I wonder how much commonality there will be for the individual handlers vs. having generic handlers like InProgress, which might end up looking at state as part of their logic.
> 
> If the key event handlers end up being generic InProgress ones, then it might make more sense to define a table with handlers key'ed by just the event, and then have the code look at the current state to determine what to do.

@eriknordmark , thanks for reviewing.

Currently it is a one-off thing, and I could place every other handler on a specific state-event transition. IMHO, we can keep it like this, and re-visit this if we end up creating too many generic handlers like this. I've renamed this handler as just "handleRestart", and got rid of the state suffix.

Reply:
Makes sense; you can see how the code evolves.